sys/amd64/amd64/zcond_machdep.c

```diff
@@ -0,0 +1,257 @@
+
+#include <vm/vm.h>
+#include <vm/pmap.h>
+#include <vm/vm_extern.h>
+#include <vm/vm_page.h>
+
+#include <machine/cpufunc.h>
+#include <machine/pmap.h>
+#include <machine/zcond.h>
+
+struct pmap zcond_pmap;
+
+void
+zcond_before_patch(void)
+{
+}
+
+void
+zcond_after_patch(void)
+{
+	mfence();
+}
+
+void
+zcond_before_rendezvous(struct zcond_md_ctxt *ctxt)
+{
+	pmap_qremove_zcond();
+	ctxt->cr3 = rcr3();
+	load_cr3(zcond_pmap.pm_cr3);
+}
+
+void
+zcond_after_rendezvous(struct zcond_md_ctxt *ctxt)
+{
+	load_cr3(ctxt->cr3);
+	invltlb();
+}
+
+static void
+insn_nop(uint8_t insn[], size_t size)
+{
+	int i;
+
+	if (size == ZCOND_INSN_SHORT_SIZE) {
+		for (i = 0; i < ZCOND_INSN_SHORT_SIZE; i++) {
+			insn[i] = nop_short_bytes[i];
+		}
+	} else {
+		for (i = 0; i < ZCOND_INSN_LONG_SIZE; i++) {
+			insn[i] = nop_long_bytes[i];
+		}
+	}
+}
+
+static void
+insn_jmp(uint8_t insn[], size_t size, vm_offset_t offset)
+{
+	int i;
+
+	if (size == ZCOND_INSN_SHORT_SIZE) {
+		insn[0] = ZCOND_JMP_SHORT_OPCODE;
+		insn[1] = offset;
+	} else {
+		insn[0] = ZCOND_JMP_LONG_OPCODE;
+		for (i = 0; i < ZCOND_INSN_LONG_SIZE - 1; i++) {
+			insn[i + 1] = (offset >> (i * 8)) & 0xFF;
+		}
+	}
+}
+
+void
+zcond_get_patch_insn(struct patch_point *p, uint8_t insn[], size_t *size)
+{
+	uint8_t *patch_addr;
+	vm_offset_t offset;
+
+	patch_addr = (uint8_t *)p->patch_addr;
+	if (*patch_addr == nop_short_bytes[0]) {
+		/* two byte nop */
+		*size = ZCOND_INSN_SHORT_SIZE;
+		goto nop;
+	} else if (*patch_addr == nop_long_bytes[0]) {
+		*size = ZCOND_INSN_LONG_SIZE;
+		goto nop;
+	} else if (*patch_addr == ZCOND_JMP_SHORT_OPCODE) {
+		/* two byte jump */
+		*size = ZCOND_INSN_SHORT_SIZE;
+		goto jmp;
+	} else if (*patch_addr == ZCOND_JMP_LONG_OPCODE) {
+		/* five byte jump */
+		*size = ZCOND_INSN_LONG_SIZE;
+		goto jmp;
+	} else {
+		panic("unexpected opcode: %02hhx", *patch_addr);
+	}
+
+nop:
+	/* replace nop with jmp */
+	offset = p->lbl_true_addr - p->patch_addr - *size;
+	insn_jmp(insn, *size, offset);
+	return;
+
+jmp:
+	/* replace jmp with nop */
+	insn_nop(insn, *size);
+}
+
+/**********************
+ * pmap functionality *
+ ***********************/
+
+static vm_offset_t zcond_patch_va;
+static pt_entry_t *zcond_patch_pte;
+
+static pt_entry_t *
+zcond_pte(vm_offset_t va)
+{
+	pml5_entry_t *pml5e;
+	pml4_entry_t *pml4e;
+	pdp_entry_t *pdpe;
+	pd_entry_t *pde;
+	pt_entry_t *pte;
+	vm_pindex_t pml5_idx, pml4_idx, pdp_idx, pd_idx;
+	vm_paddr_t mphys;
+	extern int la57;
+
+	bool is_la57 = false;
+	if (zcond_pmap.pm_type == PT_X86) {
+		is_la57 = la57;
+	}
+
+	pml4_idx = pmap_pml4e_index(va);
+	if (is_la57) {
+		pml5_idx = pmap_pml5e_index(va);
+		pml5e = &zcond_pmap.pm_pmltopu[pml5_idx];
+		KASSERT(*pml5e != 0, ("va %#jx pml5e == 0", va));
+		mphys = *pml5e & PG_FRAME;
+
+		pml4e = (pml4_entry_t *)PHYS_TO_DMAP(mphys);
+		pml4e = &pml4e[pml4_idx];
+	} else {
+		pml4e = &zcond_pmap.pm_pmltop[pml4_idx];
+	}
+
+	KASSERT(*pml4e != 0, ("va %#jx pml4e == 0", va));
```

Comment:
```suggestion
	KASSERT(*pml4e != 0, ("%s: va %#jx pml4e == 0", __func__, va));
```
Predlažem isto iz a kassertove dolje, čisto da dijagnostika bude lakša ako nešto pukne, jer trenutno je poruka ista kao i u pmap.c.